File: app/crate/observability/sentry.py
"""Sentry configuration and privacy safeguards for Crate runtimes."""

from __future__ import annotations

import copy
import os
from contextlib import contextmanager
from dataclasses import dataclass
from typing import Any
from urllib.parse import parse_qsl, urlencode, urlsplit, urlunsplit

import sentry_sdk

_FILTERED = "[Filtered]"
_DEFAULT_TRACE_SAMPLE_RATE = 0.1
_SENSITIVE_QUERY_KEYS = {
    "access_token",
    "api_key",
    "code",
    "media_ticket",
    "password",
    "refresh_token",
    "secret",
    "session",
    "signature",
    "state",
    "token",
}
_SENSITIVE_KEY_PARTS = (
    "authorization",
    "cookie",
    "credential",
    "password",
    "secret",
    "token",
)


@dataclass(frozen=True)
class SentrySettings:
    service: str
    dsn: str | None
    environment: str
    release: str | None
    traces_sample_rate: float
    profiles_sample_rate: float

    @property
    def enabled(self) -> bool:
        return bool(self.dsn)


def load_settings(service: str) -> SentrySettings:
    """Load runtime Sentry settings without requiring Sentry to be enabled."""
    dsn = os.getenv("SENTRY_DSN", "").strip() or None
    enabled = os.getenv("SENTRY_ENABLED", "true").strip().lower()
    if enabled in {"0", "false", "no", "off"}:
        dsn = None

    environment = (
        os.getenv("SENTRY_ENVIRONMENT", "").strip()
        or os.getenv("CRATE_ENV", "").strip()
        or "development"
    )
    release = os.getenv("SENTRY_RELEASE", "").strip() or None
    if release is None:
        git_sha = os.getenv("GITHUB_SHA", "").strip()
        release = f"crate-{git_sha}" if git_sha else None

    return SentrySettings(
        service=service,
        dsn=dsn,
        environment=environment,
        release=release,
        traces_sample_rate=_sample_rate(
            "SENTRY_TRACES_SAMPLE_RATE", _DEFAULT_TRACE_SAMPLE_RATE
        ),
        profiles_sample_rate=_sample_rate("SENTRY_PROFILES_SAMPLE_RATE", 0.0),
    )


def init_sentry(service: str) -> bool:
    """Initialize Sentry once for a process and attach the service tag."""
    settings = load_settings(service)
    if not settings.enabled:
        return False

    if not sentry_sdk.is_initialized():
        init_kwargs: dict[str, Any] = {
            "dsn": settings.dsn,
            "environment": settings.environment,
            "send_default_pii": False,
            "traces_sample_rate": settings.traces_sample_rate,
            "profiles_sample_rate": settings.profiles_sample_rate,
            "before_send": scrub_sentry_event,
            "before_breadcrumb": scrub_sentry_breadcrumb,
        }
        if settings.release:
            init_kwargs["release"] = settings.release
        sentry_sdk.init(**init_kwargs)

    sentry_sdk.set_tag("service", settings.service)
    return True


@contextmanager
def task_scope(task_type: str, task_id: str, queue: str):
    """Attach task metadata and timing to one worker execution."""
    if not sentry_sdk.is_initialized():
        yield None
        return

    with sentry_sdk.push_scope() as scope:
        scope.set_tag("task_type", task_type)
        scope.set_tag("queue", queue)
        scope.set_context("task", {"id": str(task_id), "type": task_type})
        with sentry_sdk.start_span(op="queue.process", name=task_type) as span:
            span.set_data("task_id", str(task_id))
            span.set_data("queue", queue)
            yield span


def scrub_sentry_event(event: dict[str, Any]) -> dict[str, Any]:
    """Return a copy of an event with credentials and PII removed."""
    return _scrub_value(copy.deepcopy(event))


def scrub_sentry_breadcrumb(
    breadcrumb: dict[str, Any], hint: Any = None
) -> dict[str, Any]:
    """Scrub breadcrumb data before it is attached to an event."""
    del hint
    return _scrub_value(copy.deepcopy(breadcrumb))


def _sample_rate(name: str, default: float) -> float:
    try:
        value = float(os.getenv(name, str(default)))
    except (TypeError, ValueError):
        value = default
    return min(1.0, max(0.0, value))


def _scrub_value(value: Any, key: str | None = None) -> Any:
    if key and _is_sensitive_key(key):
        return _FILTERED
    if key == "user" and isinstance(value, dict):
        user_id = value.get("id")
        return {"id": str(user_id)} if user_id is not None else {}
    if isinstance(value, dict):
        return {
            str(item_key): _scrub_value(item, str(item_key))
            for item_key, item in value.items()
        }
    if isinstance(value, list):
        return [_scrub_value(item) for item in value]
    if isinstance(value, tuple):
        return tuple(_scrub_value(item) for item in value)
    if isinstance(value, str):
        if key in {"url", "query_string"}:
            return _scrub_url_or_query(value, is_url=key == "url")
        return value
    return value


def _is_sensitive_key(key: str) -> bool:
    normalized = key.lower().replace("-", "_")
    return normalized in {"cookies", "set_cookie", "set_cookie_header"} or any(
        part in normalized for part in _SENSITIVE_KEY_PARTS
    )


def _scrub_url_or_query(value: str, *, is_url: bool) -> str:
    if is_url:
        parsed = urlsplit(value)
        if not parsed.query:
            return value
        return urlunsplit(
            (
                parsed.scheme,
                parsed.netloc,
                parsed.path,
                _scrub_query(parsed.query),
                parsed.fragment,
            )
        )
    return _scrub_query(value)


def _scrub_query(query: str) -> str:
    return urlencode(
        [
            (key, _FILTERED if key.lower() in _SENSITIVE_QUERY_KEYS else value)
            for key, value in parse_qsl(query, keep_blank_values=True)
        ],
        safe="[]",
    )

File: app/readplane/internal/observability/sentry.go
package observability

import (
	"net/http"
	"net/url"
	"os"
	"strconv"
	"strings"
	"time"

	"github.com/getsentry/sentry-go"
	sentryhttp "github.com/getsentry/sentry-go/http"
)

const defaultTraceSampleRate = 0.1

var sensitiveQueryKeys = map[string]struct{}{
	"access_token":  {},
	"api_key":       {},
	"code":          {},
	"media_ticket":  {},
	"password":      {},
	"refresh_token": {},
	"secret":        {},
	"session":       {},
	"signature":     {},
	"state":         {},
	"token":         {},
}

var sensitiveHeaderKeys = map[string]struct{}{
	"authorization":       {},
	"cookie":              {},
	"proxy-authorization": {},
	"x-api-key":           {},
}

// SentrySettings contains the runtime configuration for one service.
type SentrySettings struct {
	Service            string
	DSN                string
	Environment        string
	Release            string
	TracesSampleRate   float64
	ProfilesSampleRate float64
}

func (settings SentrySettings) Enabled() bool {
	return settings.DSN != ""
}

// LoadSentrySettings reads configuration without requiring Sentry to be enabled.
func LoadSentrySettings(service string) SentrySettings {
	dsn := strings.TrimSpace(os.Getenv("SENTRY_DSN"))
	if disabled(os.Getenv("SENTRY_ENABLED")) {
		dsn = ""
	}

	environment := firstNonEmpty(
		os.Getenv("SENTRY_ENVIRONMENT"),
		os.Getenv("CRATE_ENV"),
		"development",
	)
	release := strings.TrimSpace(os.Getenv("SENTRY_RELEASE"))
	if release == "" {
		release = strings.TrimSpace(os.Getenv("GITHUB_SHA"))
		if release != "" {
			release = "crate-" + release
		}
	}

	return SentrySettings{
		Service:            service,
		DSN:                dsn,
		Environment:        environment,
		Release:            release,
		TracesSampleRate:   sampleRate("SENTRY_TRACES_SAMPLE_RATE", defaultTraceSampleRate),
		ProfilesSampleRate: sampleRate("SENTRY_PROFILES_SAMPLE_RATE", 0),
	}
}

// InitSentry initializes the process-wide SDK. Observability errors never stop
// the readplane from serving traffic, so callers should log the returned error.
func InitSentry(service string) (bool, func(), error) {
	settings := LoadSentrySettings(service)
	if !settings.Enabled() {
		return false, func() {}, nil
	}

	options := sentry.ClientOptions{
		Dsn:                   settings.DSN,
		Environment:           settings.Environment,
		Release:               settings.Release,
		EnableTracing:         settings.TracesSampleRate > 0,
		TracesSampleRate:      settings.TracesSampleRate,
		SampleRate:            1,
		SendDefaultPII:        false,
		BeforeSend:            ScrubEvent,
		BeforeSendTransaction: ScrubEvent,
		MaxBreadcrumbs:        50,
	}
	if err := sentry.Init(options); err != nil {
		return false, func() {}, err
	}
	sentry.ConfigureScope(func(scope *sentry.Scope) {
		scope.SetTag("service", settings.Service)
	})

	return true, func() { sentry.Flush(2 * time.Second) }, nil
}

// WrapHTTP adds request transactions and panic recovery to a net/http handler.
func WrapHTTP(handler http.Handler) http.Handler {
	return sentryhttp.New(sentryhttp.Options{
		Repanic:         true,
		WaitForDelivery: false,
	}).Handle(handler)
}

// ScrubEvent removes credentials and user-identifying data before sending.
func ScrubEvent(event *sentry.Event, _ *sentry.EventHint) *sentry.Event {
	if event == nil {
		return nil
	}
	if event.Request != nil {
		event.Request.URL = scrubURL(event.Request.URL)
		event.Request.QueryString = scrubQuery(event.Request.QueryString)
		event.Request.Cookies = ""
		event.Request.Data = ""
		for key := range event.Request.Headers {
			if _, ok := sensitiveHeaderKeys[strings.ToLower(key)]; ok {
				event.Request.Headers[key] = "[Filtered]"
			}
		}
	}
	event.User.Email = ""
	event.User.IPAddress = ""
	event.User.Username = ""
	event.User.Name = ""
	event.User.Data = nil
	event.Extra = scrubMap(event.Extra).(map[string]interface{})
	for key, context := range event.Contexts {
		event.Contexts[key] = scrubMap(context).(map[string]interface{})
	}
	for key, value := range event.Tags {
		if sensitiveKey(key) {
			event.Tags[key] = "[Filtered]"
		} else {
			event.Tags[key] = value
		}
	}
	return event
}

func sampleRate(name string, fallback float64) float64 {
	value, err := strconv.ParseFloat(strings.TrimSpace(os.Getenv(name)), 64)
	if err != nil {
		value = fallback
	}
	if value < 0 {
		return 0
	}
	if value > 1 {
		return 1
	}
	return value
}

func disabled(value string) bool {
	switch strings.ToLower(strings.TrimSpace(value)) {
	case "0", "false", "no", "off":
		return true
	default:
		return false
	}
}

func firstNonEmpty(values ...string) string {
	for _, value := range values {
		if trimmed := strings.TrimSpace(value); trimmed != "" {
			return trimmed
		}
	}
	return ""
}

func sensitiveKey(key string) bool {
	normalized := strings.ToLower(strings.ReplaceAll(key, "-", "_"))
	if normalized == "cookies" || normalized == "set_cookie" {
		return true
	}
	for _, part := range []string{"authorization", "cookie", "credential", "password", "secret", "token"} {
		if strings.Contains(normalized, part) {
			return true
		}
	}
	return false
}

func scrubURL(raw string) string {
	parsed, err := url.Parse(raw)
	if err != nil || parsed.RawQuery == "" {
		return raw
	}
	parsed.RawQuery = scrubQuery(parsed.RawQuery)
	return parsed.String()
}

func scrubQuery(raw string) string {
	if raw == "" {
		return ""
	}
	values, err := url.ParseQuery(raw)
	if err != nil {
		return raw
	}
	for key := range values {
		if _, ok := sensitiveQueryKeys[strings.ToLower(key)]; ok {
			values.Set(key, "[Filtered]")
		}
	}
	return values.Encode()
}

func scrubMap(values map[string]interface{}) interface{} {
	if values == nil {
		return map[string]interface{}{}
	}
	result := make(map[string]interface{}, len(values))
	for key, value := range values {
		if sensitiveKey(key) {
			result[key] = "[Filtered]"
			continue
		}
		switch nested := value.(type) {
		case map[string]interface{}:
			result[key] = scrubMap(nested)
		case []interface{}:
			items := make([]interface{}, len(nested))
			for index, item := range nested {
				if child, ok := item.(map[string]interface{}); ok {
					items[index] = scrubMap(child)
				} else {
					items[index] = item
				}
			}
			result[key] = items
		default:
			result[key] = value
		}
	}
	return result
}
